File: 2nd_semester/Maths/106bombyx/B-MAT-200-BDX-2-1-106bombyx-marius.marolleau/bombyx/parser.py
from .display import print_usage


def check_args(argv):
    if len(argv) == 2 and argv[1] == "-h":
        print_usage()
        return 0

    if len(argv) == 3:
        try:
            int(argv[1])
            float(argv[2])
        except ValueError:
            return 1
        if int(argv[1]) >= 0 and 1 <= float(argv[2]) <= 4:
            return 0
        return 1
    if len(argv) == 4:
        try:
            int(argv[1])
            int(argv[2])
            int(argv[3])
        except ValueError:
            return 1
        if int(argv[1]) >= 0 and 0 <= float(argv[2]) <= float(argv[3]):
            return 0
        return 1

File: 4th_semester/OOP/Arcade/B-OOP-400-BDX-4-1-arcade-marius.marolleau/src/Parser/Parser.cpp
/*
** EPITECH PROJECT, 2024
** B-OOP-400-BDX-4-1-arcade-marius.marolleau
** File description:
** Parser
*/
#include <fstream>
#include <string>
#include <iostream>
#include <sstream>
#include <vector>
#include "../Error/Error.hpp"
#include <cstdlib>

std::string tabLib[13] = {"lib/arcade_ndk++.so",
                        "lib/arcade_aalib.so",
                        "lib/arcade_libcaca.so",
                        "lib/arcade_allegro5.so",
                        "lib/arcade_xlib.so",
                        "lib/arcade_gtk+.so",
                        "lib/arcade_sfml.so",
                        "lib/arcade_irrlicht.so",
                        "lib/arcade_opengl.so",
                        "lib/arcade_vulkan.so",
                        "lib/arcade_ncurses.so",
                        "lib/arcade_sdl2.so",
                        "lib/arcade_qt5.so"};

static 
void displayUsage(void)
{
    std::ifstream ifs("src/Parser/usage.txt");
    std::string line;

    if (!ifs.is_open())
        return;
    while (std::getline(ifs, line))
        std::cout << line << std::endl;
    return;
}

bool checkLib(char **argv)
{
    std::ifstream file(argv[1]);
    
    if (!file.is_open())
        throw ErrorParser("Can't open the library.");
    file.close();
    for (auto elem : tabLib)
        if (elem == argv[1])
            return true;
    throw ErrorParser("This library isn't authorized.");
}

bool checkEnv()
{
    char* display = std::getenv("DISPLAY");

    if (display == nullptr) {
        return false;
    }
    return true;
}

bool checkArgs(const int argc, char **argv)
{
    try {
        if (argc != 2) {
            displayUsage();
            throw ErrorParser("Bad args.");
        }
        if (!checkLib(argv))
            throw ErrorParser("Bad lib !");
        if (!checkEnv())
            throw ErrorParser("Bad env !");
    }
    catch (const ErrorParser &e) {
        std::cout << e.what() << std::endl;
        return false;
    }
    return true;
}
